### PR TITLE
json output

### LINKED ISSUE #93: JSON output support
For further processing, e.g., visualization, it would be great to have JSON output support.

The current code is tailored towards line by line output with indentation and would need some refactoring.
I started looking into it. :)

### PR DESCRIPTION
**Pull request by orangecms**
- WIP: JSON output support
- add json method to FirmwareVolume

to resolve #93

### CODE REVIEW

**Comment by theopolis** [comment]
So right now each "printable" class has a `showinfo` (I think this is consistent) and this method calls the subsequent `showinfo` on any "child like" objects it manages.

Since each of these methods is a bit custom, meaning they are not always iterating and printing a `.attrs` member, we might need a similar `to_dict` method on each that returns a dictionary of what would have been shown in `showinfo`. Then we can define a common `children` or similar key that always contains the children objects.

That `to_dict` can be fed to the JSON serializer classes.

**Comment by orangecms** [comment]
Slowly but steadily... This looks promising. :)

Now one thing - and that's why I am doing this after all: Besides the size of each entry, I would like to have its absolute position in a given firmware image in order to be able to highlight it in a graphical visualization. Do you see a simple way to obtain that?

For more context, here is an example screenshot from a view that I have already implemented for PSP data structures.



You can hover and click the respective cards, and the pane on the right will show which blocks are being occupied.

The project lives here: https://github.com/orangecms/utk-web/

To try it out: https://hostile.education/utk-web/MZ92-FS0_R16_F01.psp

**Comment by theopolis** [comment]
> Slowly but steadily... This looks promising. :)

Woot, yes indeed!

**Comment by theopolis** [comment]
> I would like to have its absolute position in a given firmware image in order to be able to highlight it in a graphical visualization. Do you see a simple way to obtain that?

There's some complexity involved due to how the parsing tries to auto-handle detecting compressed parts and similarly decompressing and continuing to parse the discovered sub objects. You might be able to shim in a common `offset` field as well as a `decompressed` field indicating the offset is relative to some compressed chunk. I think this can be implemented with minor refactors to the constructors to take the `offset`, or a refactor to the output such that every time info about a child/sub object is printed, its relative offset is included.

File: uefi_firmware/uefi.py [hunk 1]
```diff
@@ -1355,6 +1355,10 @@ def process(self):
         ]
         for block in self.blocks:
             if sguid(self.guid) in ffs_guids:
+                # FIXME: there may only be a single FFS, which is the FV body
+                # see https://uefi.org/sites/default/files/resources/PI_Spec_1_7_A_final_May1.pdf
+                # section 2.1.2
```
Comment (orangecms) [inline uefi_firmware/uefi.py:1360]: Volume 3

**Comment by orangecms** [comment]
This was a tedious thing to do, and it is not fully complete, though at least firmware volumes are now printed out, and the whole thing looks like

```json
{
  "regions": [
    {
      "type": "bios",
      "data": {
        "firmwareVolumes": [
          {
            "guid": "8c8ce578-8a3d-4f1c-9935-896185c32dd3",
            "attributes": 261887,
            "revision": 2,
            "checksum": 59054,
            "size": 131072,
            "blocks": [
              {
                "size": 32,
                "length": 4096
              }
            ],
            "ffs": [
              "snip snip"
            ]
          }
        ]
      }
    }
  ]
}
```

**Comment by theopolis** [comment]
Hey @orangecms, this looks ready for review, but I wanted to confirm with you first.

**Comment by orangecms** [comment]
Yea absolutely, please go ahead. I did go a little back and forth to figure things out, now it's ready. :)

**Review by theopolis — APPROVED** [review]

**Comment by theopolis** [comment]
Woot, looks great, thanks for this @orangecms!

**Pull request merged**